### PR TITLE
Extended consumer options to provide way to override MID.

### PR DESCRIPTION
**Pull request by mstyura**
This PR is alternative approach to resolve scenario described in #584 in a way suggested by @ibc.

@ibc could you please clarify what test units are requited which you've mentioned in your [comment](https://github.com/versatica/mediasoup/pull/584#pullrequestreview-681525657)?

### CODE REVIEW

**Comment by ibc** [comment]
I'm afraid I cannot review Rust code. Regarding tests, we have tons of test units in `mediasoup/test/` folder. This feature (custom per Consumer MID) must be tested in there.

**Comment by nazar-pc** [comment]
Code looks perfectly fine, for Rust another test case should be added to `rust/tests/integration/consumer.rs` to make sure correct MID is actually used (use one of the branches in `consume_succeeds` as an example) as well as it should be mirrored on TypeScript side I think.

File: src/Transport.ts [hunk 1]
```diff
@@ -562,15 +565,19 @@ export class Transport extends EnhancedEventEmitter
 		// Set MID.
 		if (!pipe)
 		{
-			rtpParameters.mid = `${this._nextMidForConsumers++}`;
-
-			// We use up to 8 bytes for MID (string).
-			if (this._nextMidForConsumers === 100000000)
+			rtpParameters.mid = preferredMid;
+			if (!rtpParameters.mid)
```
Comment (jmillan) [inline src/Transport.ts]: Little readability change.

```suggestion
			
			if (preferredMid)
			{
			  rtpParameters.mid = preferredMid;
			}
			else
```

File: lib/Consumer.d.ts [hunk 2]
```diff
@@ -27,6 +27,11 @@ export declare type ConsumerOptions = {
      * by itself.
      */
     paused?: boolean;
+    /**
+     * The preferred MID for consumer created. If not specified will be assigned
+     * automatically to sequentlially growing number serialized to string.
```
Comment (jmillan) [inline lib/Consumer.d.ts]:
```suggestion
     * The preferred MID for the Consumer. If not specified, a sequentially growing number will be
     * assigned.
```

File: src/Consumer.ts [hunk 3]
```diff
@@ -37,6 +37,12 @@ export type ConsumerOptions =
 	 */
 	paused?: boolean;
 
+	/**
+	 * The preferred MID for consumer created. If not specified will be assigned 
+	 * automatically to sequentlially growing number serialized to string.
```
Comment (jmillan) [inline src/Consumer.ts]:
```suggestion
          * The preferred MID for the Consumer. If not specified, a sequentially growing number will be 
          * assigned.
```

File: rust/src/router/consumer.rs [hunk 4]
```diff
@@ -76,6 +76,9 @@ pub struct ConsumerOptions {
     /// consuming endpoint even before it's ready to consume it, generating “black” video until the
     /// device requests a keyframe by itself.
     pub paused: bool,
+    /// The preferred MID for consumer created. If not specified will be assigned automatically
+    /// to sequentlially growing number serialized to string.
```
Comment (jmillan) [inline rust/src/router/consumer.rs]:
```suggestion
    /// The preferred MID for the Consumer. If not specified, a sequentially growing number will be 
    /// assigned.
```

**Comment by jmillan** [comment]
LGTM, I've just made few cosmetic change requests.

**Comment by jmillan** [comment]
Adding here too the need for adding the corresponding tests in TS.

**Review by jmillan — APPROVED** [review]

**Comment by mstyura** [comment]
@jmillan could you please suggest next steps to do to make this PR merged?

File: test/test-Consumer.js [hunk 5]
```diff
@@ -989,3 +1006,44 @@ test('Consumer emits "transportclose" if Transport is closed', async () =>
 				mapConsumerIdProducerId  : {}
 			});
 }, 2000);
+
+describe('Tests which should run on fresh worker/router/transport', () =>
```
Comment (jmillan) [inline test/test-Consumer.js]: Any test after this scenario will find the worker shut down.

In order to make this more reliable the test could verify that:

1. Creating a consumer without providing a `preferredMid` assigns certain MID (`x`), without asserting the value of `x`.
2. Creating a consumer providing `preferredMid` assigns the given MID
3. Creating a consumer without providing a `preferredMid` assigns the MID (`Number(x) + 1`)

**Comment by jmillan** [comment]
> @jmillan could you please suggest next steps to do to make this PR merged?

I have added a new comment on consumer test.

File: lib/Consumer.d.ts [hunk 6]
```diff
@@ -27,6 +27,11 @@ export declare type ConsumerOptions = {
      * by itself.
      */
     paused?: boolean;
+    /**
+     * The preferred MID for the Consumer. If not specified, a sequentially growing
+     * number will be assigned.
+     */
+    preferredMid?: string;
```
Comment (ibc) [inline lib/Consumer.d.ts]: Why preferredMid instead of mid? I mean, if given it's gonna always be used, right?

Comment (mstyura) [reply lib/Consumer.d.ts]: yes, I assumed that it should always be used when specified. I could change the name to just `mid`. I've seen there are other optional parameters prefixed with preferred, so I thought it was some naming convention of optional parameters to override some default configuration.

**Review by jmillan — APPROVED** [review]

**Comment by ibc** [comment]
Thanks for the changes. Please wait until Monday for me to do a proper review and merge&release it.

File: test/test-Consumer.js [hunk 7]
```diff
@@ -497,6 +497,41 @@ test('transport.consume() succeeds', async () =>
...
+		{
+			producerId      : audioProducer.id,
+			rtpCapabilities : consumerDeviceCapabilities
+		});
+
+	expect(audioConsumer3.rtpParameters.mid).toEqual(
+		expect.stringMatching(/^[0-9]+/));
+	expect(Number(audioConsumer1.rtpParameters.mid) + 1).toBe(
+		Number(audioConsumer3.rtpParameters.mid));
+	audioConsumer3.close();
```
Comment (ibc) [inline test/test-Consumer.js]:
```suggestion
	expect(Number(audioConsumer1.rtpParameters.mid) + 1).toBe(
		Number(audioConsumer3.rtpParameters.mid));

	audioConsumer3.close();
```

**Review by ibc — APPROVED** [review]

**Comment by ibc** [comment]
Released in mediasoup 3.7.16. Thanks!

However... a Rust test is failing: https://github.com/mstyura/mediasoup/runs/2873138895?check_suite_focus=true

@nazar-pc would you mind taking a look?

**Comment by ibc** [comment]
Sorry, ignore. That's not the mediasoup mainstream branch :)

**Comment by ibc** [comment]
Nope, of course it also happens in mainstream: https://github.com/versatica/mediasoup/runs/2873156693?check_suite_focus=true

**Comment by nazar-pc** [comment]
New Rust version was released and added a new lint. Any warnings cause CI errors, will fix and bump crates versions for publishing.

UPD: BTW it literally told you what is wrong, why, how to fix and where to read more about it :slightly_smiling_face:

**Pull request merged**